csrc/scheduler/hopper_multi_matmul.cpp

```diff
@@ -0,0 +1,1524 @@
+  //   -> swizzle X2, Y2 -> [...., X1, X2', X3, Y1, Y2', Y3]
+  //   -> merge back -> [...., X', Y']
+  //   -> mma-swizzle transformations -> loop
+  // The mma-swizzle transformations for the mma output and epilogue smem
+  // tensor are the same. In indexing, we do require {X, X'} and {Y, Y'} to be
+  // mapped in CA map, however, we currently can not handle that. So we have
+  // to do the same split and merge to the mma output without actually
+  // applying the swizzle, and this check is to detect and handle this
+  // specific case. We should remove this special handling when we fix our CA
+  // mapping.
+  using SwizzleTypeMaybeLegacy =
+      std::conditional_t<legacy, Swizzle2DType, SwizzleType>;
+  if (isPowOf2(num_gigabanks)) {
+    swizzle_domain.swizzle(SwizzleTypeMaybeLegacy::XOR, axis_of_gigarow_id, -2);
+  } else {
+    swizzle_domain.swizzle(
+        SwizzleTypeMaybeLegacy::CyclicShift, axis_of_gigarow_id, -2);
+  }
+
+  if (legacy) {
+    if (repeated_pattern_size > 1) {
+      swizzle_domain.merge(-6);
+    }
+    swizzle_domain.merge(-5);
+
+    // merge back tile_size_y
+    swizzle_domain.merge(-3);
+    swizzle_domain.merge(-2);
+  }
+
+  return swizzle_domain;
+}
+
+} // namespace
+
+void HopperMultipleMatmulScheduler::run() {
+  // Clears memory spaces on intermediate tensors, calls
+  // cache{After,Before,Fork} on inputs and outputs
+  cacheInputsAndOutputs();
+
+  // Finds matmul patterns and translates them to MmaOps, then finds tensor
+  // and dimension roles for all tensors in the fusion
+  findPatterns();
+  translatePatterns();
+  findRoles();
+
+  // Defines acw_smem/bcw_smem and acr/bcr by possibly calling cacheAfter.
+  // This also collects mma_results_
+  defineOperandCaches();
+
+  // Schedules:
+  //   - global->smem (cp.async)
+  //   - smem->register (ldmatrix)
+  //   - prologue computation in registers, including broadcast to e.g.
+  //   ab=[iM, bN, iK]
+  schedulePrologues();
+
+  // schedule mma instruction output (mma_result)
+  scheduleMmaResults();
+
+  // schedule epilogue
+  scheduleEpilogue();
+
+  // schedule splitk_sum
+  scheduleSplitKSum();
+
+  setUpInlining();
+
+  // set up circular buffering. This must come after everything up to
+  // mma_result is scheduled, since everything in the main loop will need to
+  // be rotated
+  setUpCircularBuffering();
+}
+
+void HopperMultipleMatmulScheduler::cacheInputsAndOutputs() {
+  // Make sure we don't have global memory set on intermediate tensors from
+  // fusion segmentation
+  scheduler_utils::clearMemorySpace(fusion_);
+
+  // Cache inputs
+  scheduler_utils::cacheInputs(fusion_, /*unroll=*/true);
+
+  // Cache and fork outputs
+  cached_outputs_ =
+      scheduler_utils::cacheAndForkOutputs(fusion_, /*unroll=*/true);
+}
+
+void HopperMultipleMatmulScheduler::findPatterns() {
+  patterns_ = mma_utils::findMatmulPatterns(fusion_);
+  NVF_ERROR(!patterns_.empty(), "No matmul patterns were found");
+}
```

Comment:
Stuff like this should probably just go into a base class which we subclass for each new arch. What do you think?

Reply:
Make sense. Should we make this refactor and consolidation in a separate PR so the changes are more clear.